### PR TITLE
only load direct dependencies into pom.xml

### LINKED ISSUE #2481: Sometimes adding certain library adds all transitive deps
For Cloud Libraries, adding a single library to a Maven project usually adds a single dependency in `pom.xml`. For example,
```xml
<dependency>
<groupId>com.google.cloud</groupId>
<artifactId>google-cloud-spanner</artifactId>
<version>0.25.0-beta</version>
</dependency>
```

However, adding Cloud Speech adds about 60 deps. I think it adds most, if not all, transitive deps. Adding Cloud Endpoints similarly adds a lot of deps.

BTW, adding Objectify additionally adds guava, and I'm not sure if this is needed.

### PR DESCRIPTION
**Pull request by elharo**
@chanseokoh fix #2481

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 1]
```diff
@@ -115,15 +117,16 @@ void setSiteUri(URI siteUri) {
   }
 
   public synchronized List<LibraryFile> getLibraryFiles() {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: Returning different results whether a library is resolved or not seems a bit dangerous, and counter to the model (the extension point has `<library><libraryFile/><libraryFile/>…</library>`.  If you changed `getDirectDependencies()` → `getTransitiveDependencies()` you could get rid of the resolved flag and use whether `transitiveDependencies == null` as the flag.

Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: I do think this class is do for some more rework since it was desigbned under the assumption (no longer valid) that most libraries listed all their deps in plugin.xml. resolved should probably default to false. 

That said, I'm not sure I want to change more than the minimum we need until 1.4.0 is  out. It feels like it risks introducing other bugs.

Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: That's my point: changing `getLibraryFiles()` to return the transitive dependencies is counter to the original design, and it has introduced bugs.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 2]
```diff
@@ -115,15 +117,16 @@ void setSiteUri(URI siteUri) {
...
   }
 
   /**
    * @param libraryFiles artifacts associated with this library, cannot be <code>null</code>
    */
   public synchronized void setLibraryFiles(List<LibraryFile> libraryFiles) {
     Preconditions.checkNotNull(libraryFiles);
-    this.libraryFiles = new ArrayList<>(libraryFiles);
+    this.directDependencies = new ArrayList<>(libraryFiles);
+    this.transitiveDependencies = this.directDependencies;
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: This needs some thoughts, as it blindly overwrites the transitiveDependencies. It's getting hard to track the effect, and it's not clear what the contract of this method should be. However, I think we should set `resolved = false` here. I wish everything were immutable.

Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: Immutability isn't really feasible for this class because we need to load the transitive dependencies when used. Otherwise we waste a lot of time and network bandwidth loading every dependency of every library even though the user doesn't touch it.

resolved should not be set to false here because this is only called as part of initialization of a new library object. 

I might be able to not set transitiveDependencies here and instead initialize it lazily when getLibraryFiles is called. I could even call resolveDependencies from inside getLibraryFiles though that will have the side effect of making a few more of our tests ht the network. Still that hasn't been a big problem so far, so maybe that makes sense.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 1 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: getLibraryFiles() originally returned the complete graph of dependencies. I don't think that's changed. 

Let me explore using `transitiveDependencies == null` as the flag. For sanity, I'll probably rename the two methods to 

```
getDirectDependencies
getAllDependencies
```

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 2 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: At least the code is same (I see it is just renaming) and there is no behavioral change, so I think this won't break things, but looks like we should have set `resolved = false` to be correct.

Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: Never mind. I think I misunderstood how things work here.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 3]
```diff
@@ -173,44 +171,53 @@ void setTransport(String transport) {
     this.transport = transport;
   }
   
-  public synchronized boolean isResolved() {
-    return this.resolved;
+  @VisibleForTesting
+  synchronized boolean isResolved() {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: No one calls this method except tests.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFactory.java [hunk 4]
```diff
@@ -81,9 +81,9 @@ static Library create(IConfigurationElement configurationElement) throws Library
           library.setExport(Boolean.parseBoolean(exportString));
         }
         String dependencies = configurationElement.getAttribute("dependencies"); //$NON-NLS-1$
-        if ("include".equals(dependencies)) { //$NON-NLS-1$
+        if (!"include".equals(dependencies)) { //$NON-NLS-1$
           // Load the dependencies from Maven Central later
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFactory.java]: The comment doesn't seem right, if we look at the code of `setResovled()`. The method makes `library` "final" so that it won't attempt to resolve transitive dependencies.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 5]
```diff
@@ -173,44 +171,53 @@ void setTransport(String transport) {
...
+    return transitiveDependencies != null;
   }
   
   /**
    * @param resolved true iff this library contains its complete dependency graph
    */
   synchronized void setResolved(boolean resolved) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: No one calls the method with the param value of `false`, so I think we can remove the parameter too. Maybe the method name can be changed after that.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 3 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 5 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFactory.java (hunk 4 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/LibraryFactory.java]: vestigial; removed

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java [hunk 6]
```diff
@@ -194,10 +200,49 @@ public void testAddDependencies_withDuplicates()
...
+    Element artifactId = getOnlyChild(dependency, "artifactId");
+    Assert.assertEquals("objectify", artifactId.getTextContent());
+    Element version = getOnlyChild(dependency, "version");
+    Assert.assertEquals("5.1.21", version.getTextContent());
+  }
 
   private static Document parse(InputStream in)
       throws ParserConfigurationException, IOException, SAXException {
     DocumentBuilder builder = factory.newDocumentBuilder();
+    factory.setNamespaceAware(true);
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: Why was this necessary?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 7]
```diff
@@ -173,44 +170,48 @@ void setTransport(String transport) {
...
+
   /**
-   * @param resolved true iff this library contains its complete dependency graph
+   * Returns the non-transitive dependencies of this library. Useful when a separate system such as
+   * Maven will resolve the transitive dependencies later. 
    */
-  synchronized void setResolved(boolean resolved) {
-    this.resolved = resolved;
+  public List<LibraryFile> getDirectDependencies() {
+    return new ArrayList<>(this.directDependencies);
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: `ImmutableList`?

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java [hunk 8]
```diff
@@ -114,16 +105,22 @@ void setSiteUri(URI siteUri) {
     this.siteUri = siteUri;
   }
 
-  public synchronized List<LibraryFile> getLibraryFiles() {
-    return new ArrayList<>(libraryFiles);
+  /**
+   * Returns the complete list of all transitive dependencies for this library.
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: Maybe note that this can result in network traffic and may be relatively expensive.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java (hunk 6 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries.test/src/com/google/cloud/tools/eclipse/appengine/libraries/PomTest.java]: It isn't. My mistake. reverted.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 8 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java (hunk 7 above)
Comment (elharo) [reply plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/model/Library.java]: Not a bad idea. Right now I'd prefer not to make further changes.
